src/EscapeAnalysis.jl

```diff
@@ -386,6 +386,25 @@ function escape_builtin!(::typeof(getfield), args::Vector{Any}, pc::Int, state::
     return true
 end
 
+# optimization
+# =======
+
+# similar to IR_FLAG_EFFECT_FREE
+# shifting by 5 and 6 is not used by current implementation (see comment of `ssaflags` in julia.h)
+const IR_FLAG_NO_ESCAPE     = 0x01 << 5
+
+function heap_to_stack_pass!(ir::IRCode, escapes::EscapeState)
```

Comment:
Is there any reason we want to annotate this information here rather than in `find_escapes` ?

Reply:
A reason I can think of right now is to make the design more modular and clear. The annotation itself is pretty straightforward at this moment but it may get complicated later. Separating this from the analysis pass may benefit our later implementation. But I am also 100% open to make them in one run if we find this is totally acceptable in the end.

Reply:
These annotations are tightly bound to `find_escapes` and so not sure the separation will help.

Reply:
yeah that's valid point, maybe the most reasonable option now is to implement as a post-analysis process while still in `find_escapes`. I'll do the refactoring along with the below one you just suggested tmr.